packages/dds/tree/src/typeCheck.ts

```diff
@@ -0,0 +1,206 @@
+ */
+export interface Contravariant<T> {
+	_removeCovariance?: (_: T) => void;
+}
+
+/**
+ * Constrain generic type parameters to Covariant.
+ *
+ * protected _typeCheck?: Covariant<T>;
+ *
+ * @public
+ */
+export interface Covariant<T> {
+	_removeContravariance?: T;
+}
+
+/**
+ * Constrain generic type parameters to Bivariant.
+ * Unused Generic type parameters don't constrain a type at all:
+ * Adding Bivariant does the most minimal constraint:
+ * it only prevents assignment between types when neither of the two Ts extends the
+ * other.
+ *
+ * protected _typeCheck?: Bivariant<T>;
+ *
+ * @public
+ */
+export interface Bivariant<T> {
+	/**
+	 * @see {@link Bivariant}
+	 */
+	_constrainToBivariant?(_: T): void;
+}
+
+/**
+ * Constrain generic type parameters to Invariant.
+ *
+ * protected _typeCheck?: Invariant<T>;
+ *
+ * @public
+ */
+export interface Invariant<T> extends Contravariant<T>, Covariant<T> { }
+
+/**
+ * Compile time assert that X is True.
+ * To use, simply define a type:
+ * `type _check = requireTrue<your type check>;`
+ *
+ * @public
+ */
+export type requireTrue<_X extends true> = true;
+
+/**
+ * Compile time assert that X is False.
+ * To use, simply define a type:
+ * `type _check = requireFalse<your type check>;`
+ *
+ * @public
+ */
+export type requireFalse<_X extends false> = true;
+
+/**
+ * Returns a type parameter that is true iff Source is assignable to Destination.
+ *
+ * @public
+ */
+export type isAssignableTo<Source, Destination> = isAny<Source> extends true
+	? true
+	: Source extends Destination
+	? true
+	: false;
+
+/**
+ * Returns a type parameter that is true iff Subset is a strict subset of Superset.
+ *
+ * @public
+ */
+export type isStrictSubset<Subset, Superset> = isAssignableTo<Subset, Superset> extends false
+	? false
+	: isAssignableTo<Superset, Subset> extends true
+	? false
+	: true;
+
+/**
+ * Returns a type parameter that is true iff A and B are assignable to each other, and neither is any.
+ * This is useful for checking if the output of a type meta-function is the expected type.
+ *
+ * @public
+ */
+export type areSafelyAssignable<A, B> = eitherIsAny<A, B> extends true
+	? false
+	: isAssignableTo<A, B> extends true
+	? isAssignableTo<B, A>
+	: false;
+
+/**
+ * Returns a type parameter that is true iff A is any or B is any.
+ *
+ * @public
+ */
+export type eitherIsAny<A, B> = true extends isAny<A> | isAny<B> ? true : false;
+
+/**
+ * Returns a type parameter that is true iff T is any.
+ *
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type isAny<T> = boolean extends (T extends {} ? true : false) ? true : false;
+
+/**
+ * When working with conditional types,
+ * Typescript is not very good at tracking things which have been proven along branches.
+ * This type function returns its first input only if it extends its second input.
+ * The second input should be the `TExt` of the nearest enclosing true branch of a `T extends TExt` condition.
```

Comment:
I think this one would really benefit from a short code example too

Reply:
I think Tim wrote this one? I don't actually know how its indented to be used. It doesn't have a test either. I'm just gonna delete it.